bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/target/env/GuideProbeTargets.java

```diff
@@ -194,147 +172,392 @@ public GuideProbe getGuider() {
     }
 
+    /**
+     * Iterator to all of the SPTargets in this object, including the BAGS target (which leads).
+     */
     @Override
     public Iterator<SPTarget> iterator() {
-        return targetOptions.iterator();
+        return getTargets().iterator();
+    }
+
+    public Option<SPTarget> getBagsTarget() {
+        return bagsTarget;
     }
 
     public Option<SPTarget> getPrimary() {
-        return targetOptions.getPrimary();
+        return primaryTarget;
     }
 
-    @Override
-    public GuideProbeTargets selectPrimary(Option<SPTarget> primary) {
-        return new GuideProbeTargets(guider, targetOptions.selectPrimary(primary));
+    /**
+     * Determine if the primary guide star is the guide star as selected by BAGS.
+     * If the primary guide star is not selected, this returns false.
+     * @return returns true if the primary guide star is the BAGS star, else false
+     */
+    public boolean primaryIsBagsTarget() {
+        return primaryTarget.exists(t -> bagsTarget.exists(t::equals));
     }
 
-    @Override
-    public GuideProbeTargets selectPrimary(SPTarget primary) {
-        return new GuideProbeTargets(guider, targetOptions.selectPrimary(primary));
+    /**
+     * Determine if the primary guide star is a manual target.
+     * If the primary guide star is not selected, this returns false.
+     * @return returns true if the primary guide star is a manual target, else false
+     */
+    public boolean primaryIsManualTarget() {
+        return primaryTarget.exists(manualTargets::contains);
     }
 
-    @Override
-    public GuideProbeTargets setPrimary(SPTarget primary) {
-        return new GuideProbeTargets(guider, targetOptions.setPrimary(primary));
+    /**
+     * Set the BAGS target. If the primary is the old BAGS target or unset, then the primary is set to the
+     * new BAGS target automatically; otherwise, it is maintained.
+     * @param targetOption the new BAGS target
+     * @return a new GuideProbeTargets with the primary target and BAGS target as described
+     */
+    public GuideProbeTargets setBagsTarget(final Option<SPTarget> targetOption) {
+        if (targetOption.equals(bagsTarget))
+            return this;
+
+        final Option<SPTarget> primaryTargetNew = primaryIsBagsTarget() || primaryTarget.isEmpty() ? targetOption : primaryTarget;
+        return new GuideProbeTargets(guider, targetOption, primaryTargetNew, manualTargets);
     }
 
-    @Override
-    public Option<Integer> getPrimaryIndex() {
-        return targetOptions.getPrimaryIndex();
+    /**
+     * Set the BAGS target. If the primary is the old BAGS target, then the primary is set to the
+     * new BAGS target automatically; otherwise, it is maintained.
+     * @param target the new BAGS target
+     * @return a new GuideProbeTargets with the primary target and BAGS target as described
+     */
+    public GuideProbeTargets setBagsTarget(final SPTarget target) {
```

Comment:
Somehow the `setXXX` methods that actually create a copy throw me off. Could it be clearer if they were called `withBagsTarget` then you know it is a copy with an update?